internal/pipeline/pipeline.go

```diff
@@ -79,14 +82,49 @@ func (p *Pipeline) String() string {
 	return fmt.Sprintf("%+v", *p)
 }
 
+// Validate will validate all fields in the Pipeline `p`
+// 'generators', 'transformers' and 'validators' share same schema and
+// they are valid if all functions in them are ALL valid.
+func (p *Pipeline) Validate() error {
+	if p.APIVersion != kptAPIVersion {
+		return fmt.Errorf("apiVersion %s is not valid, should be %s",
+			p.APIVersion, kptAPIVersion)
+	}
+	if p.Kind != pipelineKind {
+		return fmt.Errorf("kind %s is not valid, should be %s",
+			p.Kind, pipelineKind)
+	}
+	for _, s := range p.Sources {
+		if s == "." || s == "./*" {
+			continue
+		}
+		if err := ValidatePath(s); err != nil {
+			return fmt.Errorf("source path invalid: %w", err)
+		}
+	}
+	fns := []Function{}
+	for _, fs := range [][]Function{p.Generators, p.Transformers, p.Validators} {
+		for _, fn := range fs {
+			fns = append(fns, fn)
+		}
+	}
```

Comment:
This can be simplified by:
```Go
fns = append(fns, p.Generators...)
fns = append(fns, p.Transformers...)
fns = append(fns, p.Validators...)

```